### ISSUE TITLE
Lot of errors in logs

### ISSUE THREAD

**Issue by jlatvala** [Racailloux/home-assistant-pijuice#2]
PiJuice integration writes this kind of errors into log quite often (5897 times in less than 2 days in example below). 
However, in most times the PiJuice integration gets all the values, I guess.

Maybe some error handling should be enhanced? 
If falue_not_got, try again (or skip and try again next time), or something.

---

This error originated from a custom integration.

Logger: homeassistant.helpers.entity
Source: custom_components/pijuice/sensor.py:192
Integration: pijuice (documentation, issues)
First occurred: June 9, 2021, 16:58:55 (5897 occurrences)
Last logged: 13:23:29

Update for sensor.pijuice_io_current fails
Update for sensor.pijuice_battery_status fails
Update for sensor.pijuice_power_input_status fails
Update for sensor.pijuice_power_input_io_status fails
Update for sensor.pijuice_charge fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 350, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 558, in async_device_update
    raise exc
  File "/config/custom_components/pijuice/sensor.py", line 192, in async_update
    data = bus.read_i2c_block_data(i2c_address, self._index, self._size)
  File "/usr/local/lib/python3.8/site-packages/smbus2/smbus2.py", line 617, in read_i2c_block_data
    ioctl(self.fd, I2C_SMBUS, msg)
TimeoutError: [Errno 110] Operation timed out

**Comment by Racailloux** [comment]
Hello,
What logger does send you those errors?

Could you send me the whole log file and perhaps a description of your installation (Raspberry, GPIOs  connected/used, ...)

This seems that the library behind can't reach the I2C bus. Do you have any other elements connected in the I2C?
I will see what I can do to add some securities around the data read by I2C library

**Comment by jlatvala** [comment]
I can see the errors in the normal HA logs (Configuration / Logs).

I have enabled I2C only for this PiJuice component and no other componets use it.

My installation is Raspberry Pi 4 Model B and I have SSD disk connected to its USB3 port. I have also ConBee II connected to USB2 port. PiJuice is the only component attached to GPIO.

**Comment by jlatvala** [comment]
In addition to the errors, also accurate data is got from PiJuice, see attached image.
![pijuice](https://user-images.githubusercontent.com/25170791/121817767-fea9e680-cc8b-11eb-9db3-20ab2459a364.png)

Also, my power cord is attached into Raspberry Pi, not into PiJuice.

**Comment by Racailloux** [comment]
Ok, I am quite surprised because I got quite the same setup(R.Pi4 + SSD on USB).
Powering in from R.Pi or PiJuice is OK for I2C communication. That is one of the parameters grabbed.

What OS do you us to run Home Assistant?
Does your PiJuice Hat has some settings modified or is it "just" under factory settings?

**Comment by jlatvala** [comment]
I have just the Home Assistant Operating System installed.
And just took the PiJuice HAT out from its box and installed on top of my Pi.

Installed/enabled I2C using this add-on:
https://github.com/adamoutler/HassOSConfigurator/tree/main/Pi4EnableI2C
Rebooted the whole Pi (not just HA) twice as its logs suggest and then removed the add-on.

I don't think my HAT is not installed properly (because you just push it deep enough into GPIO pins and thats it). 

If some pins are bent a little, it shouldn't harm HAT (it should still connect).
If some pins are totally bent (not having connect at all), then PiJuice integration should not report any values, I guess.

**Comment by Racailloux** [comment]
Ok, My running installation is quite in the same configuration and this runs without trouble.
I no longer recommands HassOSConfigurator for security reasons. But that has no influence for this issue.

I think more about an hardware issue on the PiJuice. I believe that you shouldn't get a second PiJuice hardware but perhaps you have another R.Pi 4B.
If yes, can you check with the other R.PI 4B (in case of)?

In parallel, I will see own to catch those "Timeout" and "Remote I/O error".

**Comment by jlatvala** [comment]
Unfortunately I don't have second R.PI (or PiJuice :) ).

I now took PiJuice out from my R.PI and then plugged it back on. Still seeing those errors.

**Comment by vencakratky** [comment]
Please, can you help me?
HA: Home Assistant 2021.5.5
Logger: homeassistant.components.sensor
Source: custom_components/pijuice/sensor.py:191
Integration: Senzor (documentation, issues)
First occurred: 13:33:14 (9 occurrences)
Last logged: 13:33:14

pijuice: Error on device update!
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 407, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 516, in async_device_update
    await task
  File "/config/custom_components/pijuice/sensor.py", line 191, in async_update
    bus = SMBus(i2c_bus)
  File "/usr/local/lib/python3.8/site-packages/smbus2/smbus2.py", line 280, in __init__
    self.open(bus)
  File "/usr/local/lib/python3.8/site-packages/smbus2/smbus2.py", line 310, in open
    self.fd = os.open(filepath, os.O_RDWR)
FileNotFoundError: [Errno 2] No such file or directory: '/dev/i2c-1'

**Comment by Racailloux** [comment]
Hello @vencakratky.
This is a different issue that you got : You don't have the I2C activated (this is the communication port used to exchanged with PiJuice).
Please follow the [prerequisite](https://github.com/Racailloux/home-assistant-pijuice#prerequisite) to activate it.
the plugin can't activate it until this is depending on the operating system used.

**Comment by Racailloux** [comment]
@jlatvala Please test the release 1.1.0 on which I added some protection on read errors.
Instead of having a huge exception message, the integration will rise a warning message in the logs. Also the value would not be corrupted and your graphs should display "normally".

This doesn't explain why you have so much errors (I think about an hardware issue or a parasited environment?).
Please also have a try with a SD card to use standard raspbian and the official Pijuice libraries, to check if you also get thoses errors on another software environment.

**Comment by vencakratky** [comment]
> 
> 
> Hello @vencakratky.
> This is a different issue that you got : You don't have the I2C activated (this is the communication port used to exchanged with PiJuice).
> Please follow the [prerequisite](https://github.com/Racailloux/home-assistant-pijuice#prerequisite) to activate it.
> the plugin can't activate it until this is depending on the operating system used.

Sorry, it was my Mistake. A had to 2x Reboot after installing "HassOS I2C Configurator" thank you :)

**Comment by jlatvala** [comment]
I changed my manual installation to HACS installation and installed (actually updated) 1.1.0 through HACS.
Now the errors are warnings, but still getting lot of them (more than 200 in two hours):
------
This error originated from a custom integration.

Logger: custom_components.pijuice.sensor
Source: custom_components/pijuice/sensor.py:229
Integration: pijuice (documentation, issues)
First occurred: 11:52:05 (206 occurrences)
Last logged: 13:42:04

PiJuice: Error while retrieving data for sensor 'Battery voltage' (Error code: '[Errno 110] Operation timed out')
PiJuice: Error while retrieving data for sensor 'Battery current' (Error code: '[Errno 110] Operation timed out')
PiJuice: Error while retrieving data for sensor 'IO voltage' (Error code: '[Errno 110] Operation timed out')
PiJuice: Error while retrieving data for sensor 'IO current' (Error code: '[Errno 110] Operation timed out')
PiJuice: Error while retrieving data for sensor 'Battery status' (Error code: '[Errno 121] Remote I/O error')
-----


My PiJuice or R-Pi might be somehow faulty and thats the reason why it generates these errors/warnings. However, otherwise PiJuice and R.Pi works great, haven't seen any other problems.



About HACS installation...
I removed the old manual installation (removed custom_components/pijuice directory and removed pijuice-lines from configuration.yaml). 
Then I installed pijuice integration repository through HACS.
Then I tried to add pijuice integration to HA using HA - Configuration - Integrations - Add integration button, but PiJuice is not there at all.
Because of that I added the needed lines manually into configuration.yaml and then it works.
Should I create another ticket for this?

**Comment by Racailloux** [comment]
That's normal that you setup the configuration.yaml file for integration (depeding on your preferences). So that's not an issue.

**Comment by jlatvala** [comment]
Just an update...

I moved my HA from SD card to USB SSD and at the same time I moved from 32-bit to 64-bit HA.
Checked the logs 2 hours after the move and noticed that the errors/warnings mentioned in this issue are not there anymore. So problem 100% solved now :)
